doc/man3/tst_res.3
.\"
.\" $Id: tst_res.3,v 1.2 2008/06/10 05:52:02 subrata_modak Exp $
.\"
.\" Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of version 2 of the GNU General Public License as
.\" published by the Free Software Foundation.
.\"
.\" This program is distributed in the hope that it would be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\" Further, this software is distributed without any warranty that it is
.\" free of the rightful claim of any third person regarding infringement
.\" or the like.  Any license provided herein, whether implied or
.\" otherwise, applies only to this software file.  Patent licenses, if
.\" any, provided herein do not apply to combinations of this program with
.\" other software, or any other product whatsoever.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.\" Contact information: Silicon Graphics, Inc., 1600 Amphitheatre Pkwy,
.\" Mountain View, CA  94043, or:
.\"
.\" http://www.sgi.com
.\"
.\" For further information regarding this notice, see:
.\"
.\" http://oss.sgi.com/projects/GenInfo/NoticeExplan/
.\"
.TH TST_RES 3 01/21/2011 "Linux Test Project"
.SH NAME
tst_resm \- Print result message
.sp
tst_resm_hexd \- Print result message, including specified buffer in hexadecimal format
.sp
tst_brkm \- Print result message and break remaining test cases
.sp
tst_flush \- Print any messages pending because of CONDENSE mode, and flush output stream
.sp
tst_exit \- Exit test with a meaningful exit value
.sp
tst_environ \- Keep results coming to original stdout
.SH SYNOPSIS
\fB#include "test.h"\fR
.P
.P
\fBvoid tst_resm(int \fIttype\fB, char *\fItmesg, [arg ...]\fR)
.P
\fBvoid tst_resm_hexd(int \fIttype\fB, const void *\fIbuf\fB, size_t \fIsize\fB,
char *\fItmesg, [arg ...]\fR)
.P
\fBvoid tst_brkm(int \fIttype\fB, void (*\fIfunc\fB)(), char *\fItmesg,
[arg ...]\fR)
.P
\fBvoid tst_flush()
.P
\fBvoid tst_exit()
.P
\fBint  tst_environ()
.P
\fBextern int tst_count;
.br
extern int tst_range;
.br
\fR
.SH DESCRIPTION
.SS Introduction
This library of functions are used by UNICOS tests to report results to
standard output in a consistent manner.  It is assumed that tests using this
library have a distinct number of test cases, and that each test case is
distinct and uniquely identified by the test case number.  It is also assumed
that test case results are printed in consecutive order, starting with 1.
The library maintains a set of global variables (\fBTCID\fR, \fBTST_TOTAL\fR,
\fBtst_count\fR), which are used by the various functions to format the
results and to keep track of the current result reporting state (i.e. how many
total test cases there are, and how many have been reported so far) for the
calling test.
.P
The \fBTCID\fR and \fBTST_TOTAL\fR global variables are externed in the
library, and MUST be defined/initialized by tests using the library.
\fBTCID\fR must be set to the \fBT\fRest \fBC\fRase \fBID\fRentifier, and
\fBTST_TOTAL\fR must be set to the total number of test cases that will be
reported.
.P
The other global variables are available as externs to tests linked with
tst_res.o.  \fBtst_count\fR is the running count of the number of test
results that have been reported so far.  The library initializes it to 0, and
it should not be modified by the test.
The details are described below under the appropriate functions.
.SS Arguments
.RS 5
.TP 10
.I ttype
test result type; one of \fBTPASS, TFAIL, TBROK, TCONF, TWARN\fR, or
\fBTINFO\fR (explained below).
.TP 10
.I fname
pointer to a character string holding the name of a file whose contents will
be printed on a new line following \fItmesg\fR.  If this pointer is NULL, it
is ignored.
.TP 10
.I tmesg, [arg ...]
pointer to a character string containing a message explaining the test
result.  This character string can contain percent-sign conversion
specifications as in \fBprintf\fR(3C) with corresponding \fIarg\fR arguments
following the \fItmesg\fR argument.
.br
\fBNOTE:\fR These routines use static space internally to hold the
expanded message.  The maximum size allowed for an expanded message is
2048 bytes.
.TP 10
.I func
pointer to a function which performs either the cleanup necessary prior to
exiting the test or some function executed at the end of each iteration of a
loop.
.TP 10
.I buf
pointer to a buffer whose contents will be printed in hexadecimal format.
.TP 10
.I size
size of the buffer.
.RE
.SS Result Types
The possible test result types defined in \fBtest.h\fR are as follows:
.RS 5
.TP 10
.B TPASS
The test case produced expected results.
.TP 10
.B TFAIL
The test case produced unexpected results.
.TP 10
.B TBROK
A resource needed to execute the test case was not available (e.g. a
temporary file could not be opened).
.TP 10
.B TCONF
The test case was not appropriate for the current hardware or software
configuration (e.g. MLS was not enabled).
.TP 10
.B TWARN
The testing procedure caused undesirable side effects that did not affect
test results (e.g. a temporary file could not be removed after all test
results were recorded).
.TP 10
.B TINFO
An informative message about the execution of the test that does not
correspond to a test case result and does not indicate a problem.
.RE
.SS Function Descriptions

\fBtst_resm()\fR and \fBtst_resm_hexd()\fR are the basic
result reporting functions. They report 1 or more test case results of the
specified \fIttype\fR.  All result types are valid for these functions.  The
\fBtst_range\fR global variable indicates the number of results that will be
reported for each call to one of these functions.  It is initialized by the
library to 1, but may be set to a value > 1 by the test.  Each call to one of
these functions will result in \fBtst_range\fR test case results being
reported, each with an identical message (\fItmesg\fR).  \fBtst_range\fR is
always reset to 1 by the library before returning to the caller.
.P
\fBtst_brk()\fR and \fBtst_brkm()\fR are used to report results for all test
cases remaining in the test, and then call a cleanup function.  The only
result types that are valid for these functions are: \fBTFAIL, TBROK,
and TCONF\fR.  When called with a \fIttype\fR of \fBTFAIL\fR or
\fBTBROK\fR, one result of the specified \fIttype\fR will be printed,
followed by results of type \fBTBROK\fR for the remaining test cases.  When
called with a \fIttype\fR of \fBTCONF\fR, the specified
\fIttype\fR will be printed for all remaining test cases.  If \fIfunc\fR is
not NULL, \fBtst_brk()\fR and \fBtst_brkm()\fR will call \fIfunc\fR after all
results have been printed.  If the call to \fIfunc\fR returns,
\fBtst_brk()\fR and \fBtst_brkm()\fR will then call \fBtst_exit()\fR.  If
\fIfunc\fR is NULL, \fBtst_brk()\fR and \fBtst_brkm()\fR return to the caller
after all results have been printed.  If \fBtst_brk()\fR is called with a
\fIfname\fR argument, the contents of the file will only be printed for the
first reported result.  \fBtst_brk()\fR takes the \fIfname\fR argument
whereas \fBtst_brkm()\fR does not.
.P
\fBtst_flush()\fR is used to print any results pending because of
\fBCONDENSE\fR or \fBNOPASS\fR modes (described below), and flushes the
output stream.
.P
\fBtst_exit()\fR is used to allow tests to exit with a meaningful exit
value.  A bit is set in the exit value for each of the non passing test
case result types (TFAIL, TBROK, and TWARN) encountered by the library.
Thus any bit which is set in the exit value indicates that the
corresponding result flag was used at least once in the test run.
.P
The current bit fields for the result types are as follows:
.RS 5
.TP 10
TPASS
0000	/* .... .... */
.TP 10
TFAIL
0001	/* .... ...1 */
.TP 10
TBROK
0002	/* .... ..1. */
.TP 10
TWARN
0004	/* .... .1.. */
.TP 10
TINFO
0020	/* ...1 .... */
.TP 10
TCONF
0040	/* ..1. .... */
.RE
.P
NOTE: \fBTPASS and TINFO\fR do not have an effect
on the test program exit status.
.P
\fBtst_environ()\fR is used to ensure that results reported by this library
will go to the original stdout, even if the test changes the original stdout
to another file, or closes it.  A test may want to do this in order to
redirect output that normally goes to stdout (e.g. printf() output) to a
file.  \fBtst_environ()\fR returns 0 upon successful completion, and -1 if it
encountered any problems.
.SS Output Modes
Four output display modes are supported by the \fBtst_resm()\fR family of
functions to enhance output readability. The active mode is controlled via
the environment variable \fBTOUTPUT\fR, which must be set prior to the start
of the test in order to have any effect (see \fBksh\fR(1) for information on
environment variables).  The supported modes are as follows:
.RS 5
.TP 15
.B VERBOSE
A test result output line is generated for each test result.  This is the
default mode.
.TP 15
.B CONDENSE
Consecutive, identical PASS, FAIL, BROK, CONF, and RETR test results are
condensed into one output line.  The test case number field contains the range
of results involved.  WARN and INFO output lines are not condensed, but
printed as usual.
.TP 15
.B NOPASS
All PASS, CONF, INFO, and RETR output lines are discarded (i.e. not printed),
and consecutive, identical FAIL and BROK output lines are condensed as in
\fBCONDENSE\fR mode.  WARN output lines are printed as usual.
.TP 15
.B DISCARD
All output lines are discarded.
.RE
.SH EXAMPLES
.nf
#include "test.h"

char *TCID = "tsttcs01"; /* set test case identifier */
int TST_TOTAL = 15;      /* set total number of test results */

main()
{
		.
		.
	/* a successful test result */
	tst_resm(TPASS, "\fIwhat was tested\fR");
		.
		.

	/* break all remaining test results */
	tst_brkm(TBROK, cleanup, "\fIwhat didn't work\fR");
		/* or */
	tst_brk(TBROK, file, cleanup, "\fIwhat didn't work\fR");
		.
		.

	/* exit after all test results have been passed to tst_res */
	tst_exit();
}
.fi
.P
Sample output:
.RS 5
.nf
tsttcs01    1       PASS  :  Able to create MAXUP processes
tsttcs01    2       FAIL  :  Too many processes (MAXUP+1) created
tsttcs01    3       BROK  :  tabinfo(PROCTAB, &tbs) failed; errno = 13: Permission denied
.fi
.SH "SEE ALSO"
tst_setup(1),
printf(3C),
ksh(1).
.SH DIAGNOSTICS
.P
A WARN result message will be printed if any of the following occur:
.RS 5
.P
If an invalid test type is specified.
.P
If \fBtst_count\fR is negative.
.P
If one of the \fBtst_brk[m]()\fR routines is called with a test type
other than \fBTFAIL, TBROK, TCONF\fR.
.P
If there are any problems opening/reading/writing the contents of \fIfname\fR.
.RE
.SH LIMITATIONS
If \fIfname\fR is NULL and \fItmesg\fR is NULL or empty, the result message
will be empty.  This allows a test to not print a message for a result, but
it is not advised.
.SH NOTES
In multithreaded environment, output of \fBtst_resm_hexd()\fR may be interleaved
with messages produced by other threads.
.SH BUGS
.P
The programmer is free to alter the value of \fBtst_count\fR causing possible
test result order problems.


ChangeLog
LTP-20100831

commit dd59d49d75598d163f1b4599067a9f0dcb70e6c4
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Tue Aug 31 23:09:49 2010 -0700

    The cacheflush system call does not return EINVAL in the latest Linux
    kernel. see the link:
    A related patch about the cacheflush function:
    http://lkml.org/lkml/2009/4/9/203
    But Ralf had refused this patch:
    http://www.spinics.net/lists/linux-man/msg00906.html
    cacheflush01 testcase checks if the return value is EINVAL when the
    cache argument is not one of ICACHE, BCACHE or DCACHE. So it will fail
    in all boards.
    
    In this modification, checking return value SUCCESS will be added,
    instead of checking EINVAL
    
    Signed-off-by: Feng Gao <feng.gao@windriver.com>

:100644 100644 cce0b70... d0a4b1f... M	testcases/kernel/syscalls/cacheflush/cacheflush01.c

commit ebea45b92a1e30e4480f25291ef4c1ef2eeaa4bf
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Tue Aug 31 17:14:21 2010 -0700

    Get rid of unnecessary style divergence.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100755 100755 0ebeca9... b08a2d9... M	testcases/commands/at/at_allow01
:100755 100755 6bca177... 9f9f433... M	testcases/commands/at/at_deny01

commit 08640ca184c1fc1e24e966c3d806a722f3ca2154
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 20:53:18 2010 -0700

    Add more assertions.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:000000 100644 0000000... 2c590b1... A	testcases/open_posix_testsuite/conformance/interfaces/access/assertions.xml

commit 3cdde2f86af90b3fd4e2b1468a12adfd2a9523f0
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 20:52:52 2010 -0700

    Note which timer failed in output.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 ac043b0... 8e149ea... M	testcases/open_posix_testsuite/conformance/behavior/timers/1-1.c

commit 207f47c4c9b4783c1a18be32431f5a1b319b9e5a
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 19:34:34 2010 -0700

    Update some assertions for raise(2) to SUSv4 reqs.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 38d9542... aa47b7c... M	testcases/open_posix_testsuite/conformance/interfaces/raise/assertions.xml

commit 9c6d2b4cf4f7c166813ddee280ab2bf128eb521c
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 03:36:51 2010 -0700

    Fix bad syntax that I thought I hadn't committed.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100755 100755 f731d6c... 0ebeca9... M	testcases/commands/at/at_allow01

commit 414747ced26d943a22a3f767deb7f5b4658adb6e
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 03:32:18 2010 -0700

    Forgot the sleeps.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 9b10030... 6d5eb7f... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-2.c
:100644 100644 a3be617... 60e5241... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-3.c
:100644 100644 34e3afd... abc3279... M	testcases/open_posix_testsuite/conformance/interfaces/pthread_attr_setstack/2-1.c
:100755 100755 01984c5... 536d407... M	testcases/open_posix_testsuite/scripts/generate-makefiles.sh

commit 3222fbecbfb2924a6b901a8e2048c70e1a993018
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 01:22:07 2010 -0700

    Remove bogus comment.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 c7d81b5... 72a1fcd... M	testcases/open_posix_testsuite/conformance/behavior/timers/2-1.c

commit 7187f0921382bf375cd111b8a480a74ce5786945
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 01:18:23 2010 -0700

    Repeat robustness work done to 1-1 for 1-2 and 1-3.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 a9d104c... 9b10030... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-2.c
:100644 100644 af23ebe... a3be617... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-3.c

commit 6c2750a55eaa30fe708c78087688c0d468bfb807
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 01:12:07 2010 -0700

    Make test more robust.
    
    1. Catch fork() == -1 errors.
    2. Sleep so wait(2) should always succeed.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 1efd520... f46ee8f... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-1.c

commit 873f6e5b28a6a23367a7117aad436378e56ad09a
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sat Aug 28 01:42:53 2010 -0700

    Fix ltp-aiodio.sh script so it execs aio-stress.
    
    Singed-off-by: Jacky Malcles <jacky.malcles@bull.net>

:100755 100755 ed2bce4... ab54c54... M	testscripts/ltp-aiodio.sh

commit 4fd92183833eed128bb6420ece7a6dd45dcd8ec5
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sat Aug 28 01:40:51 2010 -0700

    I've prepared a patch for at_allow01 that essentially duplicates the
    fixes that were checked in for at_deny01. This fixes the script to
    actually run the test when the at utility in stead of returning TCONF,
    correcting test environment, and let the script run normally without
    changing the password. See commit
    a589d12b0b49287d10f43b1a1fb72950819588eb for more details.
    
    Signed-off-by: Henry Yei <hyei@mvista.com>

:100755 100755 cf5b97c... f731d6c... M	testcases/commands/at/at_allow01

commit 217d9b778d275bcbaf5a9fa9e1a5ac2716521cad
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sat Aug 28 01:39:12 2010 -0700

    Chain cmds together to avoid intermediate errors.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100755 100755 b94bd8f... 6bca177... M	testcases/commands/at/at_deny01

commit 9eeea7939bd73b295ae376dc045e5fbb984d0f38
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Wed Aug 25 20:40:23 2010 -0700

    Fix error message (was hardcoded to SIGALARM).
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>


lib/tst_res.c
/*
 * Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
 *    AUTHOR            : Kent Rogers (from Dave Fenner's original)
 *    CO-PILOT          : Rich Logan
 *    DATE STARTED      : 05/01/90 (rewritten 1/96)
 * Copyright (c) 2009-2016 Cyril Hrubis <chrubis@suse.cz>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it would be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * Further, this software is distributed without any warranty that it is
 * free of the rightful claim of any third person regarding infringement
 * or the like.  Any license provided herein, whether implied or
 * otherwise, applies only to this software file.  Patent licenses, if
 * any, provided herein do not apply to combinations of this program with
 * other software, or any other product whatsoever.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Contact information: Silicon Graphics, Inc., 1600 Amphitheatre Pkwy,
 * Mountain View, CA  94043, or:
 *
 * http://www.sgi.com
 *
 * For further information regarding this notice, see:
 *
 * http://oss.sgi.com/projects/GenInfo/NoticeExplan/
 */

#define _GNU_SOURCE

#include <pthread.h>
#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>

#include "test.h"
#include "usctest.h"
#include "ltp_priv.h"

long TEST_RETURN;
int TEST_ERRNO;

#define VERBOSE      1
#define NOPASS       3
#define DISCARD      4

#define MAXMESG      80		/* max length of internal messages */
#define USERMESG     2048	/* max length of user message */
#define TRUE         1
#define FALSE        0

/*
 * EXPAND_VAR_ARGS - Expand the variable portion (arg_fmt) of a result
 *                   message into the specified string.
 *
 * NOTE (garrcoop):  arg_fmt _must_ be the last element in each function
 *		     argument list that employs this.
 */
#define EXPAND_VAR_ARGS(buf, arg_fmt, buf_len) do {\
	va_list ap;				\
	assert(arg_fmt != NULL);		\
	va_start(ap, arg_fmt);			\
	vsnprintf(buf, buf_len, arg_fmt, ap);	\
	va_end(ap);				\
	assert(strlen(buf) > 0);		\
} while (0)

static pthread_mutex_t tmutex = PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP;

static void check_env(void);
static void tst_condense(int tnum, int ttype, const char *tmesg);
static void tst_print(const char *tcid, int tnum, int ttype, const char *tmesg);

static int T_exitval = 0;	/* exit value used by tst_exit() */
static int T_mode = VERBOSE;	/* flag indicating print mode: VERBOSE, */
				/* NOPASS, DISCARD */

static char Warn_mesg[MAXMESG];	/* holds warning messages */

/*
 * These are used for condensing output when NOT in verbose mode.
 */
static int Buffered = FALSE;	/* TRUE if condensed output is currently */
				/* buffered (i.e. not yet printed) */
static char *Last_tcid;		/* previous test case id */
static int Last_num;		/* previous test case number */
static int Last_type;		/* previous test result type */
static char *Last_mesg;		/* previous test result message */

int tst_count = 0;

/*
 * These globals must be defined in the test.
 */
extern char *TCID;		/* Test case identifier from the test source */
extern int TST_TOTAL;		/* Total number of test cases from the test */


struct pair {
	const char *name;
	int val;
};

#define PAIR(def) [def] = {.name = #def, .val = def},
#define STRPAIR(key, value) [key] = {.name = value, .val = key},

#define PAIR_LOOKUP(pair_arr, idx) do {                       \
	if (idx < 0 || (size_t)idx >= ARRAY_SIZE(pair_arr) || \
	    pair_arr[idx].name == NULL)                       \
		return "???";                                 \
	return pair_arr[idx].name;                            \
} while (0)

const char *strttype(int ttype)
{
	static const struct pair ttype_pairs[] = {
		PAIR(TPASS)
		PAIR(TFAIL)
		PAIR(TBROK)
		PAIR(TCONF)
		PAIR(TWARN)
		PAIR(TINFO)
	};

	PAIR_LOOKUP(ttype_pairs, TTYPE_RESULT(ttype));
}

#include "errnos.h"
#include "signame.h"

static void tst_res__(const char *file, const int lineno, int ttype,
                      const char *arg_fmt, ...)
{
	pthread_mutex_lock(&tmutex);

	char tmesg[USERMESG];
	int len = 0;
	int ttype_result = TTYPE_RESULT(ttype);

	if (file && (ttype_result != TPASS && ttype_result != TINFO))
		len = sprintf(tmesg, "%s:%d: ", file, lineno);
	EXPAND_VAR_ARGS(tmesg + len, arg_fmt, USERMESG - len);

	/*
	 * Save the test result type by ORing ttype into the current exit
	 * value (used by tst_exit()).
	 */
	T_exitval |= ttype_result;

	check_env();

	/*
	 * Set the test case number and print the results, depending on the
	 * display type.
	 */
	if (ttype_result == TWARN || ttype_result == TINFO) {
		tst_print(TCID, 0, ttype, tmesg);
	} else {
		if (tst_count < 0)
			tst_print(TCID, 0, TWARN,
				  "tst_res(): tst_count < 0 is not valid");

		/*
		 * Process each display type.
		 */
		switch (T_mode) {
		case DISCARD:
			break;
		case NOPASS:	/* filtered by tst_print() */
			tst_condense(tst_count + 1, ttype, tmesg);
			break;
		default:	/* VERBOSE */
			tst_print(TCID, tst_count + 1, ttype, tmesg);
			break;
		}

		tst_count++;
	}

	pthread_mutex_unlock(&tmutex);
}

static void tst_condense(int tnum, int ttype, const char *tmesg)
{
	int ttype_result = TTYPE_RESULT(ttype);

	/*
	 * If this result is the same as the previous result, return.
	 */
	if (Buffered == TRUE) {
		if (strcmp(Last_tcid, TCID) == 0 && Last_type == ttype_result &&
		    strcmp(Last_mesg, tmesg) == 0)
			return;

		/*
		 * This result is different from the previous result.  First,
		 * print the previous result.
		 */
		tst_print(Last_tcid, Last_num, Last_type, Last_mesg);
		free(Last_tcid);
		free(Last_mesg);
	}

	/*
	 * If a file was specified, print the current result since we have no
	 * way of retaining the file contents for comparing with future
	 * results.  Otherwise, buffer the current result info for next time.
	 */
	Last_tcid = malloc(strlen(TCID) + 1);
	strcpy(Last_tcid, TCID);
	Last_num = tnum;
	Last_type = ttype_result;
	Last_mesg = malloc(strlen(tmesg) + 1);
	strcpy(Last_mesg, tmesg);
	Buffered = TRUE;
}

void tst_flush(void)
{
	NO_NEWLIB_ASSERT("Unknown", 0);

	pthread_mutex_lock(&tmutex);

	/*
	 * Print out last line if in NOPASS mode.
	 */
	if (Buffered == TRUE && T_mode == NOPASS) {
		tst_print(Last_tcid, Last_num, Last_type, Last_mesg);
		Buffered = FALSE;
	}

	fflush(stdout);

	pthread_mutex_unlock(&tmutex);
}

static void tst_print(const char *tcid, int tnum, int ttype, const char *tmesg)
{
	int err = errno;
	const char *type;
	int ttype_result = TTYPE_RESULT(ttype);
	char message[USERMESG];
	size_t size;

	/*
	 * Save the test result type by ORing ttype into the current exit value
	 * (used by tst_exit()).  This is already done in tst_res(), but is
	 * also done here to catch internal warnings.  For internal warnings,
	 * tst_print() is called directly with a case of TWARN.
	 */
	T_exitval |= ttype_result;

	/*
	 * If output mode is DISCARD, or if the output mode is NOPASS and this
	 * result is not one of FAIL, BROK, or WARN, just return.  This check
	 * is necessary even though we check for DISCARD mode inside of
	 * tst_res(), since occasionally we get to this point without going
	 * through tst_res() (e.g. internal TWARN messages).
	 */
	if (T_mode == DISCARD || (T_mode == NOPASS && ttype_result != TFAIL &&
				  ttype_result != TBROK
				  && ttype_result != TWARN))
		return;

	/*
	 * Build the result line and print it.
	 */
	type = strttype(ttype);
	if (T_mode == VERBOSE) {
		size = snprintf(message, sizeof(message),
				"%-8s %4d  %s  :  %s", tcid, tnum, type, tmesg);
	} else {
		size = snprintf(message, sizeof(message),
				"%-8s %4d       %s  :  %s",
				tcid, tnum, type, tmesg);
	}

	if (size >= sizeof(message)) {
		printf("%s: %i: line too long\n", __func__, __LINE__);
		abort();
	}

	if (ttype & TERRNO) {
		size += snprintf(message + size, sizeof(message) - size,
				 ": errno=%s(%i): %s", tst_strerrno(err),
				 err, strerror(err));
	}

	if (size >= sizeof(message)) {
		printf("%s: %i: line too long\n", __func__, __LINE__);
		abort();
	}

	if (ttype & TTERRNO) {
		size += snprintf(message + size, sizeof(message) - size,
				 ": TEST_ERRNO=%s(%i): %s",
				 tst_strerrno(TEST_ERRNO), (int)TEST_ERRNO,
				 strerror(TEST_ERRNO));
	}

	if (size >= sizeof(message)) {
		printf("%s: %i: line too long\n", __func__, __LINE__);
		abort();
	}

	if (ttype & TRERRNO) {
		size += snprintf(message + size, sizeof(message) - size,
				 ": TEST_RETURN=%s(%i): %s",
				 tst_strerrno(TEST_RETURN), (int)TEST_RETURN,
				 strerror(TEST_RETURN));
	}

	if (size + 1 >= sizeof(message)) {
		printf("%s: %i: line too long\n", __func__, __LINE__);
		abort();
	}

	message[size] = '\n';
	message[size + 1] = '\0';

	fputs(message, stdout);
}

static void check_env(void)
{
	static int first_time = 1;
	char *value;

	if (!first_time)
		return;

	first_time = 0;

	/* BTOUTPUT not defined, use default */
	if ((value = getenv(TOUTPUT)) == NULL) {
		T_mode = VERBOSE;
		return;
	}

	if (strcmp(value, TOUT_NOPASS_S) == 0) {
		T_mode = NOPASS;
		return;
	}

	if (strcmp(value, TOUT_DISCARD_S) == 0) {
		T_mode = DISCARD;
		return;
	}

	T_mode = VERBOSE;
	return;
}

void tst_exit(void)
{
	NO_NEWLIB_ASSERT("Unknown", 0);

	pthread_mutex_lock(&tmutex);

	tst_flush();

	exit(T_exitval & ~TINFO);
}

pid_t tst_fork(void)
{
	pid_t child;

	NO_NEWLIB_ASSERT("Unknown", 0);

	tst_flush();

	child = fork();
	if (child == 0)
		T_exitval = 0;

	return child;
}

void tst_record_childstatus(void (*cleanup)(void), pid_t child)
{
	int status, ttype_result;

	NO_NEWLIB_ASSERT("Unknown", 0);

	if (waitpid(child, &status, 0) < 0)
		tst_brkm(TBROK | TERRNO, cleanup, "waitpid(%d) failed", child);

	if (WIFEXITED(status)) {
		ttype_result = WEXITSTATUS(status);
		ttype_result = TTYPE_RESULT(ttype_result);
		T_exitval |= ttype_result;

		if (ttype_result == TPASS)
			tst_resm(TINFO, "Child process returned TPASS");

		if (ttype_result & TFAIL)
			tst_resm(TINFO, "Child process returned TFAIL");

		if (ttype_result & TBROK)
			tst_resm(TINFO, "Child process returned TBROK");

		if (ttype_result & TCONF)
			tst_resm(TINFO, "Child process returned TCONF");

	} else {
		tst_brkm(TBROK, cleanup, "child process(%d) killed by "
			 "unexpected signal %s(%d)", child,
			 tst_strsig(WTERMSIG(status)), WTERMSIG(status));
	}
}

pid_t tst_vfork(void)
{
	NO_NEWLIB_ASSERT("Unknown", 0);

	tst_flush();
	return vfork();
}

/*
 * Make tst_brk reentrant so that one can call the SAFE_* macros from within
 * user-defined cleanup functions.
 */
static int tst_brk_entered = 0;

static void tst_brk__(const char *file, const int lineno, int ttype,
                      void (*func)(void), const char *arg_fmt, ...)
{
	pthread_mutex_lock(&tmutex);

	char tmesg[USERMESG];
	int ttype_result = TTYPE_RESULT(ttype);

	EXPAND_VAR_ARGS(tmesg, arg_fmt, USERMESG);

	/*
	 * Only FAIL, BROK, CONF, and RETR are supported by tst_brk().
	 */
	if (ttype_result != TFAIL && ttype_result != TBROK &&
	    ttype_result != TCONF) {
		sprintf(Warn_mesg, "%s: Invalid Type: %d. Using TBROK",
			__func__, ttype_result);
		tst_print(TCID, 0, TWARN, Warn_mesg);
		/* Keep TERRNO, TTERRNO, etc. */
		ttype = (ttype & ~ttype_result) | TBROK;
	}

	tst_res__(file, lineno, ttype, "%s", tmesg);
	if (tst_brk_entered == 0) {
		if (ttype_result == TCONF) {
			tst_res__(file, lineno, ttype,
				"Remaining cases not appropriate for "
				"configuration");
		} else if (ttype_result == TBROK) {
			tst_res__(file, lineno, TBROK,
				 "Remaining cases broken");
		}
	}

	/*
	 * If no cleanup function was specified, just return to the caller.
	 * Otherwise call the specified function.
	 */
	if (func != NULL) {
		tst_brk_entered++;
		(*func) ();
		tst_brk_entered--;
	}
	if (tst_brk_entered == 0)
		tst_exit();

	pthread_mutex_unlock(&tmutex);
}

void tst_resm_(const char *file, const int lineno, int ttype,
	const char *arg_fmt, ...)
{
	char tmesg[USERMESG];

	EXPAND_VAR_ARGS(tmesg, arg_fmt, USERMESG);

	if (tst_test)
		tst_res_(file, lineno, ttype, "%s", tmesg);
	else
		tst_res__(file, lineno, ttype, "%s", tmesg);
}

typedef void (*tst_res_func_t)(const char *file, const int lineno,
		int ttype, const char *fmt, ...);

void tst_resm_hexd_(const char *file, const int lineno, int ttype,
	const void *buf, size_t size, const char *arg_fmt, ...)
{
	char tmesg[USERMESG];
	static const size_t symb_num	= 2; /* xx */
	static const size_t size_max	= 16;
	size_t offset = strlen(tmesg);
	size_t i;
	char *pmesg = tmesg;
	tst_res_func_t res_func;

	if (tst_test)
		res_func = tst_res_;
	else
		res_func = tst_res__;

	EXPAND_VAR_ARGS(tmesg, arg_fmt, USERMESG);

	if (size > size_max || size == 0 ||
		(offset + size * (symb_num + 1)) >= USERMESG)
		res_func(file, lineno, ttype, "%s", tmesg);
	else
		pmesg += offset;

	for (i = 0; i < size; ++i) {
		/* add space before byte except first one */
		if (pmesg != tmesg)
			*(pmesg++) = ' ';

		sprintf(pmesg, "%02x", ((unsigned char *)buf)[i]);
		pmesg += symb_num;
		if ((i + 1) % size_max == 0 || i + 1 == size) {
			res_func(file, lineno, ttype, "%s", tmesg);
			pmesg = tmesg;
		}
	}
}

void tst_brkm_(const char *file, const int lineno, int ttype,
	void (*func)(void), const char *arg_fmt, ...)
{
	char tmesg[USERMESG];

	EXPAND_VAR_ARGS(tmesg, arg_fmt, USERMESG);

	if (tst_test) {
		if (func) {
			tst_brk_(file, lineno, TBROK,
			         "Non-NULL cleanup in newlib!");
		}

		tst_brk_(file, lineno, ttype, "%s", tmesg);
	} else {
		tst_brk__(file, lineno, ttype, func, "%s", tmesg);
	}

	/* Shouldn't be reached, but fixes build time warnings about noreturn. */
	abort();
}

void tst_require_root(void)
{
	NO_NEWLIB_ASSERT("Unknown", 0);

	if (geteuid() != 0)
		tst_brkm(TCONF, NULL, "Test needs to be run as root");
}
